### ISSUE TITLE
FR: equivalent to printf / sprintf

### ISSUE THREAD

**Issue by aspiers** [supercollider/supercollider#3570]
I'm having a real struggle trying to achieve the equivalent of a typical `printf` with something like `%-5.3f`, or even `%.3f`, and [I'm not the only one](http://new-supercollider-mailing-lists-forums-use-these.2681727.n2.nabble.com/asStringPrec-does-not-seem-to-work-td7630979.html).

[`String#-format`](http://doc.sccode.org/Classes/String.html#-format) only allows basic interpolation, and even with things like [`padRight`](http://doc.sccode.org/Classes/String.html#-padRight) and [`asStringPrec`](http://doc.sccode.org/Classes/Float.html#-asStringPrec) this is way too hard.  It's a must-have core feature for any language really, so it'd be great if it was added to sclang.

**Comment by brianlheim** [comment]
Yep, this is definitely something I've wanted and I've seen multiple requests for it. We should implement it and consider deprecating asStringPrec in the process.

**Comment by mkb218** [comment]
There's an undocumented `fformat` method on String which looks like it may have been intended for this. 

https://github.com/supercollider/supercollider/commit/0ff292a7ec7cad871a604654f64cf24e98e63592

**Comment by brianlheim** [comment]
@mkb218 Thanks for the effort! But it should really be done as a C++ primitive, so that we don't reinvent the wheel while introducing new bugs. There is information on writing primitives in the SC help system - http://doc.sccode.org/Guides/WritingPrimitives.html. You could also take a look at the existing string primitives for ideas; the new code will probably go there (https://github.com/supercollider/supercollider/blob/develop/lang/LangPrimSource/PyrStringPrim.cpp).

**Comment by beatboxchad** [comment]
hey @mkb218 I'm tired of not being a marketable C++ dev and this looks like an afternoon's baby step toward fixing that for me. Are you in the middle of cooking something per the above suggestion, or do you mind if I take a shot?

**Comment by mkb218** [comment]
hi @beatboxchad I will likely not have time to accomplish this one until October. Go for it.

**Comment by beatboxchad** [comment]
Just an update: I was able to understand the primitive system relatively quickly, and began coding a solution. But in a dev meeting we discussed the bigger picture of the various string formatting facilities in sclang and it was suggested that this issue was a good initial candidate for an RFC, a process introduced right around the same time.

Day job and mental health stuff has interfered with my ability to volunteer the time needed to research the RFC, but I intend to catch a break and finish this by the end of the year.

**Comment by mnvr** [comment]
Might help someone meanwhile - a sclang implementation of a sibling to `asStringPrec` that uses the "%f" specifier

```supercollider
(
// The default asStringPrec uses "%g". This method allows us to instead
// use a "%f" style output.
//
// Example:
//
//  ~asStringPrecF.(-2.08008714253083e-07) == "-0.000000208008714"
//
~asStringPrecF = { |f, prec = 14|
    var as, c, pre;
    as = f.abs.asString;
    // Pass through values that don't have an negative exponent
    if (as.containsi("e-").not) { f.asStringPrec(prec) } {
        // Otherwise split on the exponent
        c = as.split($e);
        c[0].remove($.);
        pre = if(f < 0, "-0.", "0.");
        [pre,
            "0".extend(c[1].asInteger.neg - 1, $0),
            c[0]].join[..(prec + pre.size - 1)]
    }
};

~test_asStringPrecF = {
    var t, rt;
    t = UnitTest.new;
    UnitTest.reportPasses = false;
    t.assertEquals(
        ~asStringPrecF.( 2.08008714253083e-07),
        "0.00000020800871");
    t.assertEquals(
        ~asStringPrecF.(-2.08008714253083e-07),
        "-0.00000020800871");
    t.assertEquals(
        ~asStringPrecF.(9.58003511186689e-07),
        "0.00000095800351");

    100.do {
        rt = 0.5e-3.rand2;
        t.assertFloatEquals(~asStringPrecF.(rt).asFloat, rt)
    };
 };

~test_asStringPrecF.value();
)
```

Also packaged as a method, if that's more useful:

```supercollider

+ Float {
    // The default asStringPrec uses "%g". This method allows us to instead
    // use a "%f" style output.
    //
    // Example:
    //
    //  asStringPrecF(-2.08008714253083e-07) == "-0.000000208008714"
    //
    asStringPrecF { |prec = 14|
        var as, c, pre;
        as = this.abs.asString;
        // Pass through values that don't have an negative exponent
        if (as.containsi("e-").not) { ^this.asStringPrec(prec) } {
            // Otherwise split on the exponent
            c = as.split($e);
            c[0].remove($.);
            pre = if(this < 0, "-0.", "0.");
            ^[pre,
                "0".extend(c[1].asInteger.neg - 1, $0),
                c[0]].join[..(prec + pre.size - 1)]
        }
    }
}
```
